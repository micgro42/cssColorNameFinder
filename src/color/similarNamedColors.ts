import MemoryColorListRepository from '@/dataAccess/MemoryColorListRepository';
import ColorListRepository from '@/domain/ColorListRepository';
import Color from 'colorjs.io';

const COLOR_DISTANCE_ALGORITHM = '2000'; // CMC, 76, ITP, 2000, Jz

export default function getCloseNamedColors(rgb: string): { name: string; color: Color }[] {
  const color = new Color(rgb);
  const colorRepo: ColorListRepository = new MemoryColorListRepository();
  const namedRGBColors = colorRepo.getColorList();
  const namedColors = namedRGBColors.map((namedRGB) => {
    const namedColor = new Color(`rgb(${namedRGB.red}, ${namedRGB.green}, ${namedRGB.blue})`);
    return {
      name: namedRGB.name,
      color: namedColor,
      distance: namedColor.deltaE(color, COLOR_DISTANCE_ALGORITHM) as number,
    };
  });
  namedColors.sort((a, b) => {
    if (a.distance < b.distance) {
      return -1;
    }

    if (a.distance > b.distance) {
      return 1;
    }

    return 0;
  });
  return namedColors;
}
