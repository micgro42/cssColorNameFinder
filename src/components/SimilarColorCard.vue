<template>
  <div class="similarColorCard">
    <div class="similarColorCard__colorArea">
      <div class="similarColorCard__colorHalf" :style="'background-color: ' + colorName" />
      <div
        class="similarColorCard__colorHalf"
        :style="'background-color: ' + originalColor.hex()"
      />
    </div>
    <ColorNames
      :color-name="colorName"
      :color-hex="similarColor.hex()"
      :original-hex="originalColor.hex()"
    />
    <div class="similarColorCard__comparison">
      <ColorDifferenceDetail
        section-name="Luminosity"
        :similar-value="`${Math.round(similarColor.luminosity() * 100)}%`"
        :original-value="`${Math.round(originalColor.luminosity() * 100)}%`"
      >
        <p v-if="lightnessDiff > 0">
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(lightnessDiff) }}&nbsp;percentage points lighter</b>
          than <i>{{ originalColor.hex() }}</i
          >.
        </p>
        <p v-else>
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(lightnessDiff) }}&nbsp;percentage points darker</b>
          than <i>{{ originalColor.hex() }}</i
          >.
        </p>
      </ColorDifferenceDetail>
      <ColorDifferenceDetail
        section-name="Saturation"
        :similar-value="`${Math.round(similarColor.saturationv())}%`"
        :original-value="`${Math.round(originalColor.saturationv())}%`"
      >
        <p v-if="saturationDiff > 0">
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(saturationDiff) }}&nbsp;percentage points more saturated</b>
          than <i>{{ originalColor.hex() }}</i
          >.
        </p>
        <p v-else>
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(saturationDiff) }}&nbsp;percentage points less saturated</b>
          than <i>{{ originalColor.hex() }}</i
          >.
        </p>
      </ColorDifferenceDetail>
      <ColorDifferenceDetail
        section-name="Hue"
        :similar-value="`${Math.round(similarColor.hue())}°`"
        :original-value="`${Math.round(originalColor.hue())}°`"
      >
        <p>
          <i>{{ colorName }}</i> differs by
          <b>{{ hueDiff }}&nbsp;degrees</b>
          from <i>{{ originalColor.hex() }}</i
          >.
        </p>
      </ColorDifferenceDetail>
    </div>
  </div>
</template>

<script lang="ts">
import ColorDifferenceDetail from '@/components/ColorDifferenceDetail.vue';
import ColorNames from '@/components/ColorNames.vue';
import Color from 'color';
import { defineComponent, PropType } from 'vue';

export default defineComponent({
  components: { ColorDifferenceDetail, ColorNames },
  props: {
    colorName: {
      type: String,
      required: true,
    },
    colorParts: {
      type: Array as PropType<number[]>,
      required: true,
    },
    originalParts: {
      type: Array as PropType<number[]>,
      required: true,
    },
  },
  computed: {
    similarColor(): Color {
      return Color.rgb(this.colorParts);
    },
    originalColor(): Color {
      return Color.rgb(this.originalParts);
    },
    lightnessDiff(): number {
      return (
        Math.round(this.similarColor.luminosity() * 100) -
        Math.round(this.originalColor.luminosity() * 100)
      );
    },
    saturationDiff(): number {
      return (
        Math.round(this.similarColor.saturationv()) - Math.round(this.originalColor.saturationv())
      );
    },
    hueDiff(): number {
      const diff = Math.round(Math.abs(this.similarColor.hue() - this.originalColor.hue()));
      if (diff > 180) {
        return 360 - diff;
      }
      return diff;
    },
  },
});
</script>

<style lang="scss">
.similarColorCard {
  margin-top: 1rem;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background-color: white;
}

.similarColorCard__colorArea {
  min-height: 10vh;
  display: flex;
}

.similarColorCard__colorHalf {
  width: 50%;
}
</style>
