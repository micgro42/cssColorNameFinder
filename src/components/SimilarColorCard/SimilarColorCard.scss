.similarColorCard {
  margin-top: 1rem;
  padding: 1rem;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background-color: white;
}

.similarColorCard__header {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
}

.similarColorCard__colorArea {
  min-height: 10vh;
  border: 1px solid grey;
}
