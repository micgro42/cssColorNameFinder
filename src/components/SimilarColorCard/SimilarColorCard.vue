<template>
    <div class="similarColorCard">
        <div class="similarColorCard__header">
            <h3>{{ colorName }}</h3> <p>Distance: {{ roundedDistance }}</p>
        </div>
        <div
                class="similarColorCard__colorArea"
                :style="'background-color: ' + colorName"
        ></div>
    </div>
</template>

<script lang="ts">

interface SimilarColorCardThis {
    distance: number;
    colorName: string;
    roundedDistance: () => number
}

export default {
    props: {
        colorName: {
            type: String,
            required: true,
        },
        distance: {
            type: Number,
            required: true,
        },
    },
    computed: {
        roundedDistance() {
            return Math.round((this as SimilarColorCardThis).distance);
        }
    }
};
</script>

<style scoped lang="scss">
    .similarColorCard {
        margin-top: 1rem;
        padding: 1rem;
        box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
        background-color: white;
    }

    .similarColorCard__header {
        display: flex;
        justify-content: space-between;
        align-items: baseline;
    }

    .similarColorCard__colorArea {
        min-height: 10vh;
        border: 1px solid grey;
    }
</style>
