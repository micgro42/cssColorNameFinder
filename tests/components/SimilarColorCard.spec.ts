import { getColorData } from '@/color/similarNamedColors';
import { shallowMount } from '@vue/test-utils';
import SimilarColorCard from '../../src/components/SimilarColorCard.vue';

describe('SimilarColorCard', () => {
  it('calculates differences in colors correctly', () => {
    const wrapper = shallowMount(SimilarColorCard, {
      props: {
        colorName: 'red',
        similarColor: { name: 'red', ...getColorData('red') },
        originalColor: getColorData('chocolate'),
      },
    });
    expect(wrapper.vm.hueDiff).toBe(14.86);
    expect(wrapper.vm.chromaDiff).toBe(37.18);
    expect(wrapper.vm.lightnessDiff).toBe(-2.34);
  });

  it('The difference in hue is always less than 180', () => {
    const wrapper = shallowMount(SimilarColorCard, {
      props: {
        colorName: 'red',
        similarColor: { name: 'red', ...getColorData('red') },
        originalColor: getColorData('deeppink'),
      },
    });
    expect(wrapper.vm.hueDiff).toBe(43.68);
  });
});
