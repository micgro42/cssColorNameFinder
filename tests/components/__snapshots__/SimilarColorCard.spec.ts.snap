// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimilarColorCard matches snapshot 1`] = `
<div
  class="similarColorCard"
>
  <div
    class="similarColorCard__header"
  >
    <h3>
      red
    </h3>
    <p>
      Distance: 7
    </p>
  </div>
  <div
    class="similarColorCard__colorArea"
    style="background-color: red;"
  />
</div>
`;
